Grammar Changes for Safe Navigation Operator (@W-7963330)

sfsholden opened the pull request:
### What does this PR do?
1. Updates grammar rules for [safe navigator](https://salesforce.quip.com/RlRUA1eQaHFH) around: trigger-context-declaration, soql-colon-vars, switch-statement, invocation-expression, element-access-expression, member-access-expression.
2. Test changes

### What does this PR fix or reference?
@W-7963330@

### Switch Changes
#### Functionality Before
![SyntaxHighlighting_Switch_Before](https://user-images.githubusercontent.com/45604118/93363703-d8dad880-f804-11ea-8c2e-c260e4efbdfd.gif)
#### Functionality After
![SyntaxHighlighting_Switch_after](https://user-images.githubusercontent.com/45604118/93363758-eb551200-f804-11ea-88d0-018a33d3865f.gif)

### Member Access Changes
#### Functionality Before
![SyntaxHighlighting_MemberAccess_Before](https://user-images.githubusercontent.com/45604118/93364017-45ee6e00-f805-11ea-94e6-e5c2fe15c81d.gif)
#### Functionality After
![SyntaxHighlighting_MemberAccess_After](https://user-images.githubusercontent.com/45604118/93364101-5f8fb580-f805-11ea-87fa-e9243087f183.gif)

### SOQL Bindings
#### Functionality Before
![SyntaxHighlighting_SOQL_Before](https://user-images.githubusercontent.com/45604118/93364341-aa113200-f805-11ea-870c-25aa3cce1586.gif)
#### Functionality After
![SyntaxHighlighting_SOQL_After](https://user-images.githubusercontent.com/45604118/93364160-73d3b280-f805-11ea-94c4-c51f6c547254.gif)

sfsholden commented on grammars/apex.tmLanguage.cson:
```diff
@@ -889,18 +896,25 @@ repository:
       "0":
         name: "keyword.operator.conditional.colon.apex"
-    end: "(?![_[:alnum:]]|\\(|(\\?)?\\[|<)"
+    end: "(?![_[:alnum:]]|\\(|\\[|<)"
```
I really want to point this out - I had to remove this piece because it was interfering with the capturing that happened in the next pattern selection. Is this okay?

sfsholden commented on src/apex.tmLanguage.yml:
```diff
@@ -458,13 +461,16 @@ repository:
     begin: (\:)\s* # preceding colon operator
     beginCaptures:
       '0': { name: keyword.operator.conditional.colon.apex }
-    end: (?![_[:alnum:]]|\(|(\?)?\[|<)  # next character is not alpha-numeric, nor a (, [, or <. Also, test for ?[
```
I really want to point this out - I had to remove this piece because it was interfering with the capturing that happened in the next pattern selection. Is this okay?

sfsholden replied:
Resolving - added this piece back in. I updated this piece thinking it was the root issue, but it was not.

xyc commented on grammars/apex.tmLanguage.cson:
```diff
@@ -631,10 +631,17 @@ repository:
         match: "\\b(isExecuting|isInsert|isUpdate|isDelete|isBefore|isAfter|isUndelete|new|newMap|old|oldMap|size)\\b"
       }
       {
-        match: "(?:(\\.))([[:alpha:]]+)(?=\\()"
```
@lcampos Are we still actively supporting Atom?

lcampos replied:
Yes, we want other people in the community building for other tools to use this repo's tmlanguage files.

xyc commented:
@sfsholden Could we add what has been fixed here? I was trying out the test cases here. Looks like safe navigator only makes a difference in switch statements, triggers and soql bindings. Could you confirm if that's the case?

Also, do we have another story tracking integrating with VS Code?

xyc commented:
It's an edge case, (and I think what we are fixing here is more important) and this (ternery expression that looks like a safe navigator) is not correctly syntax highlighted as of now `Decimal d = b?.34:0;`, but I wonder if this PR fixes that

sfsholden commented:
> It's an edge case, (and I think what we are fixing here is more important) and this (ternery expression that looks like a safe navigator) is not correctly syntax highlighted as of now `Decimal d = b?.34:0;`, but I wonder if this PR fixes that

I'll double check to make sure. Sounds like a good test to add too. Thanks for pointing it out 👍

sfsholden commented:
> @sfsholden Could we add what has been fixed here? I was trying out the test cases here. Looks like safe navigator only makes a difference in switch statements, triggers and soql bindings. Could you confirm if that's the case?
> 
> Also, do we have another story tracking integrating with VS Code?

I added some recordings to the description from what I found and fixed. The follow-up story has not been created yet. That will happen either as part of this story, or a child story once this is approved.

sfsholden commented:
> > It's an edge case, (and I think what we are fixing here is more important) and this (ternery expression that looks like a safe navigator) is not correctly syntax highlighted as of now `Decimal d = b?.34:0;`, but I wonder if this PR fixes that
> 
> I'll double check to make sure. Sounds like a good test to add too. Thanks for pointing it out 👍

Could you actually elaborate here? I'm not seeing an issue with the syntax highlighting, it looks correct to me.

xyc commented:
> I added some recordings to the description from what I found and fixed.

Thank you. It might be helpful to mention or add to contributing docs about: "Developer: Inspect Editor Tokens and Scopes" since it would be useful information to contributors.

xyc commented:
> Could you actually elaborate here? I'm not seeing an issue with the syntax highlighting, it looks correct to me.

This is what I saw from production extension (49.9.0)

<img width="390" alt="Screen Shot 2020-09-16 at 12 29 25 PM" src="https://user-images.githubusercontent.com/679275/93383294-46443480-f818-11ea-8e5c-82a398020cc5.png">

sfsholden commented on README.md:
```diff
@@ -19,6 +19,11 @@ To **build and test** install Node.js do the following:
 
 Output grammars are output in the `grammars\` dirctory.
 
+To see the token changes from within the Salesforce VS Code Extensions:
```
@xyc Added these documentation changes. Let me know what you think.

xyc replied:
Looking good, thanks for adding the docs for contributors! cc @ekapner

xyc replied:
@sfsholden We can go ahead with merging this PR. We can always get back and improve the docs

xyc commented on README.md:
```diff
@@ -19,6 +19,11 @@ To **build and test** install Node.js do the following:
 
+To see the token changes from within the Salesforce VS Code Extensions:
+
+1. Copy the `apex.tmLanguage` results into `../salesforcedx-vscode/packages/salesforcedx-vscode-apex/node_modules/@salesforce/apex-tmlanguage/grammars/apex.tmLanguage`.
```
```suggestion
1. Copy the `apex.tmLanguage` results into `salesforcedx-vscode/packages/salesforcedx-vscode-apex/node_modules/@salesforce/apex-tmlanguage/grammars/apex.tmLanguage`.
```

xyc replied:
I suppose you mean the dev extension here. It might be helpful to also mention the contributing page for VS Code extension, or mention that for testing you could copy it to a production extension installed in https://vscode-docs.readthedocs.io/en/stable/extensions/install-extension/#your-extensions-folder.

sfsholden replied:
> > Could you actually elaborate here? I'm not seeing an issue with the syntax highlighting, it looks correct to me.
> 
> This is what I saw from production extension (49.9.0)
> 
> <img alt="Screen Shot 2020-09-16 at 12 29 25 PM" width="390" src="https://user-images.githubusercontent.com/679275/93383294-46443480-f818-11ea-8e5c-82a398020cc5.png">

Xiaoyi and I synced on this. We were seeing different colorization results, but the same tokens. I added bug W-8095488 to address this.

sfsholden commented:
> > Could you actually elaborate here? I'm not seeing an issue with the syntax highlighting, it looks correct to me.
> 
> This is what I saw from production extension (49.9.0)
> 
> <img alt="Screen Shot 2020-09-16 at 12 29 25 PM" width="390" src="https://user-images.githubusercontent.com/679275/93383294-46443480-f818-11ea-8e5c-82a398020cc5.png">

Xiaoyi and I synced on this piece. We were seeing different colorization results, but the tokens themselves were the same. I created bug W-8095488 to address it.
